### PR TITLE
KAFKA-16324: Move BrokerApiVersionsCommand to tools

### PR DESCRIPTION
**Pull request by FrankYang0529**
* Migrate `BrokerApiVersionsCommand` to tools module.
* Migrate `BrokerApiVersionsCommandTest` to user `ClusterTestExtensions`.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Review by chia7712 — COMMENTED** [review]
@FrankYang0529 could you please update the scripts too?

https://github.com/apache/kafka/blob/trunk/bin/kafka-broker-api-versions.sh#L17
https://github.com/apache/kafka/blob/trunk/bin/windows/kafka-broker-api-versions.bat#L17

File: tools/src/main/java/org/apache/kafka/tools/BrokerApiVersionsCommand.java [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+            return future;
+        }
+
+        public void awaitBrokers() {
+            List<Node> nodes;
+            do {
+                nodes = findAllBrokers();
+                if (nodes.isEmpty()) {
+                    try {
+                        Thread.sleep(50);
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/BrokerApiVersionsCommand.java]: `TimeUnit.MILLISECONDS.sleep(50);`

File: tools/src/main/java/org/apache/kafka/tools/BrokerApiVersionsCommand.java [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+                    } catch (InterruptedException e) {
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/BrokerApiVersionsCommand.java]: not sure why we need to catch it here. could you please throw it directly as it is a command-line tool

File: tools/src/main/java/org/apache/kafka/tools/BrokerApiVersionsCommand.java [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+                .withClientSslSupport()
+                .withClientSaslSupport();
+
+        private volatile boolean running = true;
+        private final Time time;
+        private final ConsumerNetworkClient client;
+        private final List<Node> bootstrapBrokers;
+        private final ConcurrentLinkedQueue<RequestFuture<ClientResponse>> pendingFutures = new ConcurrentLinkedQueue<>();
+
+        static class AdminConfig extends AbstractConfig {
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/BrokerApiVersionsCommand.java]: Could you please remove it as we can use `AbstractConfig` directly? for instance:
```java
        static AdminClient create(Map<String, ?> props) {
            return create(new AbstractConfig(ADMIN_CONFIG_DEF, props, false));
        }
```

File: tools/src/main/java/org/apache/kafka/tools/BrokerApiVersionsCommand.java [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+            AdminConfig(Map<?, ?> originals) {
+                super(ADMIN_CONFIG_DEF, originals, false);
+            }
+        }
+
+        static AdminClient create(Properties props) {
+            return create(Utils.propsToMap(props));
+        }
+
+        static AdminClient create(Map<String, ?> props) {
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/BrokerApiVersionsCommand.java]: Do we need those constructors?

File: tools/src/main/java/org/apache/kafka/tools/BrokerApiVersionsCommand.java [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+                    pendingFutures.clear();
+                }
+            }, "admin-client-network-thread");
+            networkThread.setDaemon(true);
+            networkThread.start();
+        }
+
+        private AbstractResponse send(Node target, AbstractRequest.Builder<?> request) throws InterruptedException {
+            RequestFuture<ClientResponse> future = client.send(target, request);
+            pendingFutures.add(future);
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/BrokerApiVersionsCommand.java]: Do we need another thread to poll client? Maybe we can poll the client by the same thread to simplify this tool?

**Review by chia7712 — COMMENTED** [review]
@FrankYang0529 thanks for this nice migration

**Comment by FrankYang0529** [comment]
Hi @chia7712, thanks for review. I addressed all comments. Failed cases in CI can't be reproduced on my laptop. Could you take a look when you have time? Thank you.

File: tools/src/main/java/org/apache/kafka/tools/BrokerApiVersionsCommand.java [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+
+        AdminClient(Time time, ConsumerNetworkClient client, List<Node> bootstrapBrokers) {
+            this.time = time;
+            this.client = client;
+            this.bootstrapBrokers = bootstrapBrokers;
+        }
+
+        private AbstractResponse send(Node target, AbstractRequest.Builder<?> request) {
+            RequestFuture<ClientResponse> future = client.send(target, request);
+            while (!future.isDone()) {
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/BrokerApiVersionsCommand.java:221]: Should we handle the timeout manually?

File: tools/src/main/java/org/apache/kafka/tools/BrokerApiVersionsCommand.java [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+                    LOGGER.debug("Request {} failed against node {}", request.apiKey(), broker, e);
+                }
+            }
+            throw new RuntimeException("Request " + request.apiKey() + " failed on brokers " + bootstrapBrokers);
+        }
+
+        private KafkaFuture<NodeApiVersions> getNodeApiVersions(Node node) {
+            final KafkaFutureImpl<NodeApiVersions> future = new KafkaFutureImpl<>();
+            try {
+                ApiVersionsResponse response = (ApiVersionsResponse) send(node, new ApiVersionsRequest.Builder());
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/BrokerApiVersionsCommand.java:247]: That will send the API_VERSION request to brokers one by one. Could you please leverage the `Future` (or `CompletableFuture`) to speedup it?

**Review by chia7712 — COMMENTED** [review]
@FrankYang0529 thanks for updates. a couple of comments are left.

File: tools/src/main/java/org/apache/kafka/tools/BrokerApiVersionsCommand.java (hunk 7 above)
Comment (FrankYang0529) [reply tools/src/main/java/org/apache/kafka/tools/BrokerApiVersionsCommand.java:247]: Hi @chia7712, original class sends request one by one. I think we can keep the behavior and do some refactor in the future if someone complains about this. Thanks.

https://github.com/apache/kafka/blob/c5e91546721b6b160732667ca7450d22927cc3b3/core/src/main/scala/kafka/admin/BrokerApiVersionsCommand.scala#L136-L138

File: tools/src/main/java/org/apache/kafka/tools/BrokerApiVersionsCommand.java (hunk 6 above)
Comment (FrankYang0529) [reply tools/src/main/java/org/apache/kafka/tools/BrokerApiVersionsCommand.java:221]: I add a test case `testAdminSendNoBlock` to confirm `client.send` doesn't block forever. Thank you.

**Review by chia7712 — COMMENTED** [review]
@FrankYang0529 Could you please update the class path of following scripts?

https://github.com/apache/kafka/blob/trunk/bin/kafka-broker-api-versions.sh
https://github.com/apache/kafka/blob/trunk/bin/windows/kafka-broker-api-versions.bat

**Comment by FrankYang0529** [comment]
> @FrankYang0529 Could you please update the class path of following scripts?
> 
> https://github.com/apache/kafka/blob/trunk/bin/kafka-broker-api-versions.sh https://github.com/apache/kafka/blob/trunk/bin/windows/kafka-broker-api-versions.bat

Thank you. Updated it.

File: tools/src/test/java/org/apache/kafka/tools/BrokerApiVersionsCommandTest.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.tools;
+
+//import kafka.server.KafkaBroker;
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/BrokerApiVersionsCommandTest.java]: could you please remove this?

**Review by chia7712 — COMMENTED** [review]
@FrankYang0529 thanks for updates. one tiny comment is left

File: tools/src/test/java/org/apache/kafka/tools/BrokerApiVersionsCommandTest.java (hunk 8 above)
Comment (FrankYang0529) [reply tools/src/test/java/org/apache/kafka/tools/BrokerApiVersionsCommandTest.java]: Sorry for that. Updated it.

**Review by chia7712 — APPROVED** [review]
LGTM

**Pull request merged**